### ISSUE TITLE
luacheck showing a number of warnings.

### ISSUE THREAD

**Issue by rsjaffe** [pkulchenko/serpent#26]
Running luacheck static checker on serpent resulted in a number of warnings, including a lot of reuse of variable names, shadowing the definition in the broader scope. These aren't necessarily bugs in the current version, but do make it easier to make programming mistakes in the future. I don't know if you want to fix these or not--thought I'd list them here for your information.

    serpent.lua:11:7: shadowing definition of variable 'v' on line 1
    serpent.lua:13:9: shadowing definition of variable 'v' on line 1
    serpent.lua:19:17: unused variable 'metatostring'
    serpent.lua:25:14: shadowing upvalue 's' on line 15
    serpent.lua:26:26: shadowing upvalue 's' on line 15
    serpent.lua:29:26: shadowing upvalue 's' on line 15
    serpent.lua:30:26: shadowing upvalue 's' on line 15
    serpent.lua:32:33: shadowing upvalue 'name' on line 16
    serpent.lua:33:11: shadowing upvalue 'n' on line 1
    serpent.lua:37:91: unused argument 'o'
    serpent.lua:37:94: shadowing upvalue 'n' on line 1
    serpent.lua:39:27: shadowing upvalue 'd' on line 2
    serpent.lua:44:26: shadowing upvalue argument 't' on line 15
    serpent.lua:44:29: shadowing upvalue 'name' on line 16
    serpent.lua:44:35: shadowing upvalue 'indent' on line 16
    serpent.lua:45:18: variable 'level' was previously defined as an argument on line 44
    serpent.lua:71:15: shadowing upvalue 'n' on line 1
    serpent.lua:75:13: shadowing upvalue 'sparse' on line 17
    serpent.lua:76:11: shadowing upvalue 'n' on line 1
    serpent.lua:77:29: shadowing definition of argument 'plainindex' on line 44
    serpent.lua:82:36: empty if branch
    serpent.lua:86:19: shadowing definition of variable 'sname' on line 46
    serpent.lua:89:17: shadowing definition of argument 'path' on line 44
    serpent.lua:125:30: unused argument 'k'
    serpent.lua:126:27: unused argument 't'
    serpent.lua:126:29: unused variable length argument
    serpent.lua:135:44: shadowing upvalue 'v' on line 1

**Comment by pkulchenko** [comment]
@rsjaffe, thank you for the report, but I think most of these are false positives. Here is the report that the default analyzer (lua-inspect-based) in ZeroBrane Studio generates:

```
serpent.lua:11: local variable 'v' masks earlier declaration on line 1
serpent.lua:19: unused local variable 'metatostring'; consider removing or replacing with '_'
serpent.lua:37: unused parameter 'o' in anonymous function
serpent.lua:125: unused parameter 'k' in anonymous function
serpent.lua:126: unused parameter 't' in anonymous function
```

Some of these can definitely be fixed, but the last three are not used simply because I wanted to show the actual parameters in the function declaration.

It seems like most of the reported warning fall into two categories: (1) reported parameters of functions and (2) re-declared variables in nested blocks:

```
local function s(t, opts)
  -- skipped some lines...
  local function gensym(val) return '_'..(tostring(tostring(val)):gsub("[^%w]",""):gsub("(%d%w+)",
    function(s) if not syms[s] then symn = symn+1; syms[s] = symn end return tostring(syms[s]) end)) end
```

Usage of `s` on the last line is reported as `shadowing upvalue 's' on line 15`, but this doesn't seem to be correct as these are valid usages of `s` parameter in the function and none should be reported.

```
local n, v = "serpent", "0.30" -- (C) 2012-17 Paul Kulchenko; MIT License
-- some lines skipped
for k,v in pairs(G) do globals[v] = k end -- build func to name mapping
for _,g in ipairs({'coroutine', 'debug', 'io', 'math', 'string', 'table', 'os'}) do
  for k,v in pairs(type(G[g]) == 'table' and G[g] or {}) do globals[v] = g..'.'..k end end
```

The both usages of `v` look valid to me as it's not really "shadowing" as their scope is limited to the `for` loop in both cases and in the second one it's in the nested block, which has no impact on the outer variable scope.

In my opinion, the only "risky" case of shadowing is:

```
local a = 2
...
local a = 3
```

as it changes the value of `a` down and doesn't restore it to the original value. In all other cases (doing this in a nested block or a for loop where it's automatically scoped) doesn't create a problem.

@mpeterv, FYI ^^. This may already be configurable, but I'm not sure.

**Comment by pkulchenko** [comment]
I pushed an update that fixes `metatostring` warning (33b12f1c), but the rest are either false positives or by design, as I described earlier. I'm closing this ticket, but let me know if you find any other issues or disagree.
